### PR TITLE
KEYCLOAK-17290 SAML Client - Generate AttributeConsumingService SP metadata section

### PR DESCRIPTION
**Pull request by lscorcia**
Sometimes, SAML clients need to specify a set of attributes to retrieve from the remote IdP. In SAML this is achieved via using the AttributeConsumingServiceIndex attribute of the AuthnRequest element.

It is a simple numeric index that should correspond to an attribute set declared in the SAML metadata document. Keycloak does not support defining such attribute sets in metadata yet, but adding a setting to manipulate the AuthnRequest attribute would be useful in all of the scenarios where metadata document are edited by hand.

This PR adds the ability to specify the optional Attribute Consuming Service Index in AuthnRequests. The PR is really simple as the plumbing is already in place, it just exposes the attribute to the UI.

### CODE REVIEW

**Comment by lscorcia** [comment]
It just occurred to me that it would be quite easy to fully implement the AttributeConsumingService section of the SAML SP metadata. I'm returning this PR to draft status while I complete the required work.

**Comment by lscorcia** [comment]
It was pretty easy indeed. This PR now adds the following two parameters:

immagine

They are used to build the AttributeConsumingService descriptor in the SP metadata as follows:

immagine

The attributes are automatically loaded from the mapper configured for the Identity Provider.

**Comment by hmlnarik** [comment]
@lscorcia Thank you for the contribution and apologies it took this long sitting in the PR queue. Could you please rebase the PR?

**Comment by lscorcia** [comment]
Sure! No problems for the delay, I had to deal with other priorities lately too. I really hope I can resume working on a couple of other ideas soon :( Anyway, thanks for reviewing this!

**Comment by lscorcia** [comment]
Added a couple of tests to verify the metadata document generation.

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java
Comment (mhajas) [inline services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:413]: This is specific code only for UserAttributeMapper. Maybe it should be part of mapper code itself. 

I have an idea but I am not sure this makes sense @hmlnarik could you confirm, whether it makes sense?

We could have some interface like `SamlMapperWithRequestedAttribute` with only method:
```
RequestedAttributeType getRequestedAttribute();
```

Then we could have something like this in the code to generate `attributeConsumingService`.

```
realm.getIdentityProviderMappersByAlias(getConfig().getAlias())
                    .stream()
                    .filter(mapper -> mapper instanceof ProtocolMapper)
                    .map(SamlMapperWithRequestedAttribute.class::cast)
                    .map(SamlMapperWithRequestedAttribute::getRequestedAttribute)
                    .forEach(attributeConsumingService::addRequestedAttribute);
```

WDYT?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlSpDescriptorTest.java
Comment (mhajas) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlSpDescriptorTest.java:83]: This change should be reverted after the test, so that we won't interfere other tests. You can add something like this:

```
String mapperId = identityProviderResource.getMappers()
                    .stream()
                    .filter(mapper -> "attribute-mapper-email".equals(mapper.getName()))
                    .findFirst()
                    .map(IdentityProviderMapperRepresentation::getId)
                    .orElse(null);
getCleanup().addCleanup((Runnable) () -> identityProviderResource.delete(mapperId));
```

**Review by mhajas — COMMENTED** [review]
Thank you for the Pr @lscorcia! Nice work. I added few comments in code, could you please have a look?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlSpDescriptorTest.java
Comment (lscorcia) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlSpDescriptorTest.java:83]: Uhm, I tried adding the code you suggested, but I get a 404 Not Found while running the cleanup routine:

```
12:59:04,570 INFO  [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] [KcSamlSpDescriptorTest] testAttributeConsumingServiceMappersInSpMetadata() STARTED
12:59:04,783 DEBUG [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] loading test realms
12:59:04,785 INFO  [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] importing test realms
12:59:04,786 DEBUG [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] --importing realm: consumer
12:59:05,423 DEBUG [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] --importing realm: provider
12:59:05,692 DEBUG [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] creating user for realm provider
12:59:05,909 DEBUG [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] adding identity provider to realm consumer
12:59:05,999 DEBUG [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] adding client https://localhost:8543/auth/realms/consumer to realm provider
12:59:06,022 DEBUG [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] adding client http://localhost:8280/sales-post/ to realm consumer
12:59:06,062 DEBUG [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] adding client http://localhost:8280/sales-post/.dot/ted to realm consumer
12:59:06,086 DEBUG [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] adding client broker-app to realm consumer
12:59:06,851 INFO  [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] calling all TestCleanup
12:59:06,863 ERROR [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] failed cleanup!
javax.ws.rs.NotFoundException: HTTP 404 Not Found
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.handleErrorStatus(ClientInvocation.java:247)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.extractors.DefaultEntityExtractorFactory$3.extractEntity(DefaultEntityExtractorFactory.java:50)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invokeSync(ClientInvoker.java:154)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invoke(ClientInvoker.java:115)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientProxy.invoke(ClientProxy.java:76)
	at com.sun.proxy.$Proxy149.delete(Unknown Source)
	at org.keycloak.testsuite.broker.KcSamlSpDescriptorTest.lambda$testAttributeConsumingServiceMappersInSpMetadata$1(KcSamlSpDescriptorTest.java:91)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
	at org.keycloak.testsuite.util.TestCleanup.executeCleanup(TestCleanup.java:130)
	at org.keycloak.testsuite.AbstractKeycloakTest.afterAbstractKeycloakTest(AbstractKeycloakTest.java:218)
...
```

I also tried to move the call to retrieve the mapperId inside the cleanup routine, but I get the same 404 when retrieving the mapper ID:

```
javax.ws.rs.NotFoundException: HTTP 404 Not Found
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.handleErrorStatus(ClientInvocation.java:247)
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.extractResult(ClientInvocation.java:217)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.extractors.BodyEntityExtractor.extractEntity(BodyEntityExtractor.java:64)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invokeSync(ClientInvoker.java:154)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invoke(ClientInvoker.java:115)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientProxy.invoke(ClientProxy.java:76)
	at com.sun.proxy.$Proxy149.getMappers(Unknown Source)
	at org.keycloak.testsuite.broker.KcSamlSpDescriptorTest.lambda$testAttributeConsumingServiceMappersInSpMetadata$1(KcSamlSpDescriptorTest.java:86)
```

Any idea why?

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java
Comment (lscorcia) [reply services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:413]: Sure, we can use an interface to retrieve the requested attribute and implement it in the relevant mappers. Let's wait on @hmlnarik feedback too.

Comment (hmlnarik) [reply services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:413]: Having this interface extensible to other mappers would be great. In that case though, would it be useful to keep the descriptor alteration local to the mappers? This would require allowing mappers to directly alter the `entityDescriptor` rather than supporting a single use case with adding a requested attribute while in the descriptor generation. In code terms, that would roughly mean replacing the lines 399-438 with a loop over all `IdentityProviderMapperModel`s that are instances of a (say) `SamlMetadataDescriptorUpdater` interface. This interface would contain method `updateMetadata(EntityDescriptorType descriptor)` called from this place.

Comment (mhajas) [reply services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:413]: @hmlnarik This sounds good, however, I realized there will be a problem with that. `EntityDescriptorType` is not accessible in `server-spi[-private]` modules. This means it won't be possible to implement this interface in custom/user-provided mappers. Is it a problem?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlSpDescriptorTest.java
Comment (mhajas) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlSpDescriptorTest.java:83]: Sorry for the confusion, I should have checked this. The cleanup is not necessary because the realm is removed after each test, that is why you are seeing 404.

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java
Comment (hmlnarik) [reply services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:413]: Yes, it is possible to provide a custom IdP mapper. Could you elaborate why it would be problematic if it used `services` as a dependency and module import rather than e.g. `server-spi-private`? `server-spi-private` and `services` have the same level of support, and there this interface would not be part of `server-spi` as it would be a non-established feature

Comment (mhajas) [reply services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:413]: You are right. It is the same.

**Comment by hmlnarik** [comment]
https://master-jenkins.redhat.com/job/universal-test-pipeline-server/570/

**Comment by lscorcia** [comment]
Rebased after merge of #8111 . I will now start working towards the interface idea you mentioned in the discussion above. Thank you for your insights!

**Comment by lscorcia** [comment]
Hello @mhajas and @hmlnarik ! I am trying to follow your suggestions above and add support for metadata manipulation directly in the attribute mappers. I added a commit to better show the approach I am following, will squash after discussions and review.

One thing still isn't clear to me: from the discussion, it seems like you were planning to add the attribute modification directly to `IdentityProviderMapperModel` , but that would mean that the mappers should use "well-known" configuration keys to represent the `RequestedAttribute` properties (friendly name and name). That doesn't seem too clean to me, so I'm probably missing something obvious.

In my code I actually implemented the interface into the attribute mapper itself (right now just `UserAttributeMapper`, others will follow if the approach is ok), which has the drawback that the `SAMLIdentityProvider` now has to instantiate the mappers in order to apply the metadata logic, but on the plus side this requires no property hardcoding. It also makes easier for mapper types to do different things if necessary, or even expand the configuration to handle more than one `AttributeConsumingService` in the future. What do you think?

File: themes/src/main/resources/theme/base/admin/resources/partials/realm-identity-provider-saml.html
Comment (mhajas) [inline themes/src/main/resources/theme/base/admin/resources/partials/realm-identity-provider-saml.html:312]: Should this be in this PR? Seems it was already merged as part of https://github.com/keycloak/keycloak/pull/8111

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java
Comment (mhajas) [inline services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:425]: Can we break this lambda into more calls of `map` and `filter`?

**Review by mhajas — CHANGES_REQUESTED** [review]
@lscorcia Thank you for the update and sorry for the confusion. I believe the commit you added exactly matches my idea.

However, it looks like this PR contains changes that should not. Could you please remove all changes that were already merged as part of https://github.com/keycloak/keycloak/pull/8111?

File: themes/src/main/resources/theme/base/admin/resources/partials/realm-identity-provider-saml.html
Comment (lscorcia) [reply themes/src/main/resources/theme/base/admin/resources/partials/realm-identity-provider-saml.html:312]: Hi @mhajas , this refers to the Attribute Consuming Service Name, which is different from the Attribute Consuming Service Index merged as part of #8111. I think I rebased the PR correctly, maybe squashing could help?

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java
Comment (lscorcia) [reply services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:425]: I'm not really fluent with Java streams, but I tried and couldn't make it work because the `updateMetadata` call requires both the interface instance and the mapper model:

```
realm.getIdentityProviderMappersByAliasStream(getConfig().getAlias())
  .map(mapper -> (IdentityProviderMapper) session.getKeycloakSessionFactory().getProviderFactory(IdentityProviderMapper.class, mapper.getIdentityProviderMapper()))
  .filter(target -> target instanceof SamlMetadataDescriptorUpdater)
  .map(target -> (SamlMetadataDescriptorUpdater)target)
  .forEach(metadataAttrProvider -> metadataAttrProvider.updateMetadata(mapper, entityDescriptor)); << error here, mapper is unknown  
```

I could probably change the last map call to return a pair-like object, but it seems much harder to read than the original foreach. WDYT?

Comment (mhajas) [reply services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:425]: I see. I think you are right, this way it is easier to read. I just don't like that the lambda is quite long, but I don't see a better option.

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java
Comment (lscorcia) [reply services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java]: Now that the PR uses an interface to handle this, support for more mappers is easy to add. If you don't mind however I'd keep those in separate PRs - this one has been edited too many times already.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlAttributeConsumingServiceIndexTest.java
Comment (mhajas) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlAttributeConsumingServiceIndexTest.java:17]: nitpick: Changes in this file seem redundant.

**Review by mhajas — APPROVED** [review]
Thank you for the contribution @lscorcia. The changes look good to me. I just found one last nitpick, might be worth fixing, but it is not necessary, so I am approving. @hmlnarik Could you have a look please?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlAttributeConsumingServiceIndexTest.java
Comment (lscorcia) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlAttributeConsumingServiceIndexTest.java:17]: If you're ok with that, I can fix it in the next PR I have prepared already, i.e. metadata support for "SAML Attribute to Role" mapper.

**Review by hmlnarik — APPROVED** [review]
Approving, based on @mhajas's approval

**Pull request merged**
